Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp

```diff
@@ -244,32 +262,50 @@ void DrawingAreaWC::updateRendering()
 
 void DrawingAreaWC::sendUpdateAC()
 {
-    m_rootLayer->flushCompositingStateForThisLayerOnly();
-
-    // Because our view-relative overlay root layer is not attached to the FrameView's GraphicsLayer tree, we need to flush it manually.
-    if (m_viewOverlayRootLayer) {
-        FloatRect visibleRect({ }, m_webPage.size());
-        m_viewOverlayRootLayer->flushCompositingState(visibleRect);
-    }
-
-    m_updateInfo.rootLayer = m_rootLayer->primaryLayerID();
-
-    m_commitQueue->dispatch([this, weakThis = WeakPtr(*this), stateID = m_backingStoreStateID, updateInfo = std::exchange(m_updateInfo, { })]() mutable {
-        flushLayerImageBuffers(updateInfo);
-        RunLoop::main().dispatch([this, weakThis = WTFMove(weakThis), stateID, updateInfo = WTFMove(updateInfo)]() mutable {
-            if (!weakThis)
-                return;
-            m_remoteWCLayerTreeHostProxy->update(WTFMove(updateInfo), [this, weakThis = WTFMove(weakThis), stateID](std::optional<UpdateInfo> updateInfo) {
+    bool didProcessMainFrame = false;
+    for (auto it = m_rootLayers.begin(); it != m_rootLayers.end(); it++) {
+        auto& rootLayer = *it;
+        auto frame = WebProcess::singleton().webFrame(rootLayer.frameID);
+        ASSERT(frame);
+        m_updateInfo.remoteContextHostedIdentifier = frame->layerHostingContextIdentifier();
+        m_updateInfo.rootLayer = rootLayer.layer->primaryLayerID();
+
+        bool isLastFrame = (it + 1) == m_rootLayers.end();
+        bool isMainFrame = frame->isMainFrame();
+        didProcessMainFrame = didProcessMainFrame || isMainFrame;
+        bool willCallDisplayDidRefresh = isMainFrame || (!didProcessMainFrame && isLastFrame);
+        IntSize size;
+        if (isMainFrame)
+            size = m_webPage.size();
+        else
+            size = frame->size();
+        rootLayer.layer->setSize(size);
+        rootLayer.layer->flushCompositingStateForThisLayerOnly();
+
+        // Because our view-relative overlay root layer is not attached to the FrameView's GraphicsLayer tree, we need to flush it manually.
+        FloatRect visibleRect({ }, size);
+        if (rootLayer.viewOverlayRootLayer)
+            rootLayer.viewOverlayRootLayer->flushCompositingState(visibleRect);
+
+        m_commitQueue->dispatch([this, weakThis = WeakPtr(*this), stateID = m_backingStoreStateID, updateInfo = std::exchange(m_updateInfo, { }), willCallDisplayDidRefresh]() mutable {
+            flushLayerImageBuffers(updateInfo);
```

Comment:
We should probably check weakThis before using this.

Reply:
Fixed.

Reply:
WeakPtr isn't thread safe. I should nothing for the WeakPtr in another thread.

Reply:
Reverted the fix.